### PR TITLE
added sparse.linalg norms

### PR DESCRIPTION
**Pull request by saswatpp**
Added sparse norms from https://github.com/scipy/scipy/blob/v1.4.1/scipy/sparse/linalg/_norm.py

### CODE REVIEW

**Comment by asi1024** [comment]
Thank you for the PR! Could you add unit tests?

**Comment by saswatpp** [comment]
Working on it

**Comment by toslunar** [comment]
`tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py` would have tests of `sparse.linalg.norm`.  Follow the style of `TestLsqr` in the file.

On the matrices to be tested, the testcases for `cupy.linalg.norm` (in https://github.com/cupy/cupy/blob/v8.0.0b3/tests/cupy_tests/linalg_tests/test_norms.py#L27-L59) or for `scipy.sparse.linalg.norm` (in https://github.com/scipy/scipy/blob/master/scipy/sparse/linalg/tests/test_norm.py) may be helpful.

File: tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py [hunk 1]
```diff
@@ -55,3 +55,22 @@ def test_ndarray(self, xp, sp):
         b = xp.array(self.b, dtype=self.dtype)
         x = sp.linalg.lsqr(A, b)
         return x[0]
+
+
+
+@testing.parameterize(*testing.product({
+    'ord': [None , -numpy.Inf, -2, -1, 0, 1, 2, 3, numpy.Inf],
+    'dtype': [numpy.float64, numpy.float64],
```
Comment (toslunar) [inline tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py:63]:
```suggestion
    'dtype': [numpy.float32, numpy.float64],
```
and `numpy.complex64, numpy.complex128` maybe.

File: cupyx/scipy/sparse/linalg/_norm.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        is 1-D) or a matrix norm (when `x` is 2-D) is returned.
+    Returns
+    -------
+    n : float or ndarray
+    Notes
+    -----
+    Some of the ord are not implemented because some associated functions like, 
+    _multi_svd_norm, are not yet available for sparse matrix. 
+    This docstring is modified based on numpy.linalg.norm. 
+    https://github.com/numpy/numpy/blob/master/numpy/linalg/linalg.py
```
Comment (toslunar) [inline cupyx/scipy/sparse/linalg/_norm.py:43]: - Empty lines are required before section headers (in NumPy Style).
- (Optional in this PR) Follow Google Style (see e.g. https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) like most of the other docstrings in CuPy.

Comment (saswatpp) [reply cupyx/scipy/sparse/linalg/_norm.py:42]: Please Review

File: tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py (hunk 1 above)
Comment (saswatpp) [reply tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py]: changed tests working fine

File: tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py [hunk 3]
```diff
@@ -55,3 +55,78 @@ def test_ndarray(self, xp, sp):
...
+@testing.parameterize(*testing.product({
+    'ord': [None, -numpy.Inf, -2, -1, 1, 2, numpy.Inf, 'fro'],
+    'dtype': [
+        numpy.float32,
+        numpy.float64,
+        numpy.complex64,
+        numpy.complex128
+    ],
+    'Transpose': [True, False],
+    # _min_or_max() doesn't support axes for -1, (-1, ), -2, (-2, )
```
Comment (toslunar) [inline tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py:115]: TODO: Fix it after #3474 is resolved.

File: tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py [hunk 4]
```diff
@@ -55,3 +55,78 @@ def test_ndarray(self, xp, sp):
...
+
+@testing.parameterize(*testing.product({
+    'ord': [None, -numpy.Inf, -2, -1, 1, 2, numpy.Inf, 'fro'],
+    'dtype': [
+        numpy.float32,
+        numpy.float64,
+        numpy.complex64,
+        numpy.complex128
+    ],
+    'Transpose': [True, False],
```
Comment (toslunar) [inline tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py:114]: Could you use lowercase (`transpose`) for consistency?

**Comment by toslunar** [comment]
And could you add the function to `docs/source/reference/sparse.rst`?  Then the docstring will be rendered by Sphinx.

File: tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py [hunk 5]
```diff
@@ -55,3 +55,78 @@ def test_ndarray(self, xp, sp):
         b = xp.array(self.b, dtype=self.dtype)
         x = sp.linalg.lsqr(A, b)
         return x[0]
+
+
+@testing.parameterize(*testing.product({
+    'ord': [None, -numpy.Inf, -2, -1, 1, 2, 3, numpy.Inf, 'fro'],
```
[the remark this answers is not in the document]
Comment (saswatpp) [reply tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py:61]: ord = 0 not included in scipy.sparse.

[the remark this answers is not in the document]
Comment (saswatpp) [reply tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py:61]: so both of their errors match now

File: tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py [hunk 6]
```diff
@@ -55,3 +55,78 @@ def test_ndarray(self, xp, sp):
...
+                                 accept_error=True)
+    def test_matrix_norm_axis_3(self, xp, sp):
+        a = xp.arange(9, dtype=self.dtype) - 4
+        b = a.reshape((3, 3))
+        b = sp.csr_matrix(b, dtype=self.dtype)
+        return xp.array(sp.linalg.norm(b.T, ord=self.ord, axis=(1, 0)))
+
+
+@testing.parameterize(*testing.product({
+    'ord': [None, -numpy.Inf, -2, -1, 1, 2, numpy.Inf, 'fro'],
```
[the remark this answers is not in the document]
Comment (saswatpp) [reply tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py:107]: done

File: tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (saswatpp) [reply tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py:61]: For `ord ==0`, scipy.sparse calls `(x != 0)`, now for cupyx.scipy.sparse.csr_matrix, 
```python
 def __ne__(self, other):
>       raise NotImplementedError
E       AssertionError: Only cupy raises error
```

File: tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py (hunk 4 above)
Comment (saswatpp) [reply tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py:114]: okay

File: tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py (hunk 3 above)
Comment (saswatpp) [reply tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py:115]: okay

File: tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (toslunar) [reply tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py:61]: Then could you add xfail tests?  Comment "requires `spmatrix.__ne__`" there.

BTW, I'd write `x.getnnz(axis)` but it's not been supported by CuPy yet, either.

**Comment by saswatpp** [comment]
Please Review

File: tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (saswatpp) [reply tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py]: done

**Comment by toslunar** [comment]
Jenkins, test this please.

File: tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py [hunk 7]
```diff
@@ -55,3 +55,87 @@ def test_ndarray(self, xp, sp):
...
+    def test_vector_norm(self, xp, sp):
+        a = xp.arange(9, dtype=self.dtype) - 4
+        b = a.reshape((3, 3))
+        b = sp.csr_matrix(b, dtype=self.dtype)
+        if self.transpose:
+            b = b.T
+        return sp.linalg.norm(b, ord=self.ord, axis=self.axis)
+
+    # requires spmatrix.__ne__
+    def test_ord_zero(self):
```
[the remark this answers is not in the document]
Comment (saswatpp) [reply tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py:134]: how should I do that ?

[the remark this answers is not in the document]
Comment (toslunar) [reply tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py:144]: I suppose either one of the fixes below would work.

1: How about inserting the following just before `return` in `test_vector_norm`?
```
        if xp is cupy and ord == 0:
            with self.assertRaises(NotImplementedError):
                sp.linalg.norm(b, ord=0, axis=self.axis)
            pytest.skip('requires spmatrix.__ne__')
```

2: If you'd like to keep the tests separated, could you
- turn the xfail test into a new class (e.g. `TestVectorNormXfail`) and
- repeat parameters except for `ord`?

File: tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py [hunk 8]
```diff
@@ -55,3 +55,97 @@ def test_ndarray(self, xp, sp):
...
+        numpy.complex128
+    ],
+}))
+@unittest.skipUnless(scipy_available, 'requires scipy')
+@testing.gpu
+class TestMatrixNorm(unittest.TestCase):
+
+    @testing.numpy_cupy_allclose(rtol=1e-3, atol=1e-4, sp_name='sp', 
+                                 type_check=False,
+                                 accept_error=(ValueError,
```
Comment (toslunar) [inline tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py:75]: Could you remove trailing spaces?  (See the result of Travis CI.)
```suggestion
                                 accept_error=(ValueError,
```

**Comment by cjnolet** [comment]
We could really use this on RAPIDS cuML. Any chance we could get this into the 8.0 release? @saswatpp do you think you'll have time to finish the review changes? I can help if needed.

**Comment by saswatpp** [comment]
Yes sure. just a few changes in test will be done only.

File: tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py (hunk 8 above)
Comment (saswatpp) [reply tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py]: done

**Comment by toslunar** [comment]
Jenkins, test this please.

File: cupyx/scipy/sparse/linalg/_norm.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+            M = abs(x).sum(axis=a)
+        elif ord in (2, None):
+            M = cupy.sqrt(abs(x).power(2).sum(axis=a))
+        else:
+            try:
+                ord + 1
+            except TypeError:
+                raise ValueError('Invalid norm order for vectors.')
+            M = cupy.power(abs(x).power(ord).sum(axis=a), 1 / ord)
+        return M.ravel()
```
Comment (toslunar) [inline cupyx/scipy/sparse/linalg/_norm.py:111]: Could you fix the error `'coo_matrix' object has no attribute 'ravel'`?

File: cupyx/scipy/sparse/linalg/_norm.py [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+
+__all__ = ['norm']
+
+
+def _sparse_frobenius_norm(x):
+    if cupy.issubdtype(x.dtype, cupy.complexfloating):
+        sqnorm = abs(x).power(2).sum()
+    else:
+        sqnorm = x.power(2).sum()
+    return cupy.sqrt(sqnorm).item()
```
[the remark this answers is not in the document]
Comment (toslunar) [reply cupyx/scipy/sparse/linalg/_norm.py:14]: CuPy already has the issue in sparse argmin/argmax (#3640).

File: cupyx/scipy/sparse/linalg/_norm.py (hunk 9 above)
Comment (toslunar) [reply cupyx/scipy/sparse/linalg/_norm.py:111]: Maybe the error is caused by the conflict with #3536.  Shall I fix it?

Comment (saswatpp) [reply cupyx/scipy/sparse/linalg/_norm.py:111]: how to fix it ?

Comment (toslunar) [reply cupyx/scipy/sparse/linalg/_norm.py:111]: `return M.A.ravel()` (as in [scipy/sparse/linalg/_norm.py#L180](https://github.com/scipy/scipy/blob/b89e2806b0ecd48e128767ab60973fecc69f1f1c/scipy/sparse/linalg/_norm.py#L180)) looks good, while I'm not sure.

**Comment by toslunar** [comment]
@saswatpp Can I takeover the PR?

**Comment by cjnolet** [comment]
@toslunar Is there any way this feature could be done in time to be backported for the 8.0 release? I can jump in and help as well, if needed.

**Comment by kmaehashi** [comment]
Taken over to #4017

**Pull request closed without merging**